src/levanter/models/llama.py

```diff
@@ -0,0 +1,279 @@
+        q = self.q_proj(x)  # TODO: rearrange and possibly rename
+        k = self.k_proj(x)
+        v = self.v_proj(x)
+
+        cos, sin = self.rotary_emb(v, seq_len=self.config.seq_len)
+
+        q, k = _apply_rotary_pos_emb(q, k, cos, sin, position_ids)
+
+        scale = jax.lax.rsqrt(float(self.config.HeadSize.size))
+
+        attn_weights = hax.dot("head_size", q, k) * scale
+        attn_weights = attn_weights + mask
+
+        # upcast attention to fp32. This is default for Llama Attention
+        attn_weights = attn_weights.astype(jnp.float32)
+
+        attn_weights = hnn.softmax(attn_weights, axis="key_position").astype(q.dtype)
+        attn_output = hax.dot("key_position", attn_weights, v)
+
+        attn_output = self.o_proj(attn_output)
+        return attn_output
+
+
+def _get_rotary_emb(config: LlamaConfig, head_dim: int, max_position_embeddings: int) -> LlamaRotaryEmbedding:
+    if config.rope_scaling is None:
+        return LlamaRotaryEmbedding(head_dim, max_position_embeddings)
+    else:
+        scaling_type = config.rope_scaling["type"]
+        scaling_factor = config.rope_scaling["factor"]
+        if scaling_type == "linear":
+            return LlamaLinearScalingRotaryEmbedding(head_dim, max_position_embeddings, scaling_factor=scaling_factor)
+        elif scaling_type == "dynamic":
+            return LlamaDynamicNTKScalingRotaryEmbedding(
+                head_dim, max_position_embeddings, scaling_factor=scaling_factor
+            )
+        else:
+            raise ValueError(f"Unknown RoPE scaling type {scaling_type}")
+
+
+def _rotate_half(x: NamedArray) -> NamedArray:
+    """Rotates half of the hidden dims of the input and concatenates them.
+    This is difficult to Haliax, so we do it in regular array ops.
+    """
+    x_array = x.array
+    x1 = x_array[..., : x_array.shape[-1] // 2]
+    x2 = x_array[..., x_array.shape[-1] // 2 :]
+    output = jnp.concatenate((-x2, x1), axis=-1)
+    return hax.named(output, x.axes)
+
+
+def _apply_rotary_pos_emb(
+    q: NamedArray,  # [batch, seq_len, heads, head_size]
+    k: NamedArray,  # [batch, seq_len, kv_heads, head_size]
+    cos: jnp.ndarray,  # [1, 1, seq_len, head_size]
+    sin: jnp.ndarray,  # [1, 1, seq_len, head_size]
+    position_ids: jnp.ndarray,  # [bs, seq_len]
+) -> Tuple[NamedArray, NamedArray]:
+    """Applies rotary position embedding to q and k."""
+    cos = jnp.squeeze(jnp.squeeze(cos, axis=1), axis=0)  # from [1, 1, seq_len, dim] to [seq_len, dim]
+    sin = jnp.squeeze(jnp.squeeze(sin, axis=1), axis=0)
+    cos = cos[position_ids]  # [batch, seq_len, head_size]
+    sin = sin[position_ids]  # [batch, seq_len, head_size]
+    cos = hax.named(cos, ("batch", "position", "head_size"))
+    sin = hax.named(sin, ("batch", "position", "head_size"))
+    q_embed = hax.multiply(q, cos) + hax.multiply(_rotate_half(q), sin)
+    k_embed = hax.multiply(k, cos) + hax.multiply(_rotate_half(k), sin)
```

Comment:
once cos and sin are named, replace the whole function with

q_embed = (q * cos + _rotate_half(q) * sin)
k_embed = (k * cos + _rotate_half(k) * sin)

if you still use position ids, then do 

cos = cos[Pos, position_ids]
sin = sin[Pos, position_ids]

first